app/src/main/java/at/roteskreuz/stopcorona/model/repositories/ReportingRepository.kt

```diff
@@ -170,37 +171,40 @@ class ReportingRepositoryImpl(
                 ?: throw InvalidConfigurationException.ConfigurationNotPresent
             val uploadKeysDays = configuration.uploadKeysDays
                 ?: throw InvalidConfigurationException.NullNumberOfDaysToUpload
-            val thresholdTimeFromConfiguration = ZonedDateTime.now()
+            val uploadStartIntervalNumberFromConfig = ZonedDateTime.now()
                 .minusDays(uploadKeysDays.toLong())
-                .startOfTheUtcDay()
                 .toRollingStartIntervalNumber()
 
             val dateWithMissingExposureKeys = dateWithMissingExposureKeys
 
             if (dateWithMissingExposureKeys != null) {
                 // Report only the exposure keys that have not been uploaded in the day of the previous submission.
                 val keysToUpload = teks.filter {
-                    it.rollingStartIntervalNumber >= dateWithMissingExposureKeys.startOfTheUtcDay().toRollingStartIntervalNumber() &&
-                        it.rollingStartIntervalNumber <= dateWithMissingExposureKeys.endOfTheUtcDay().toRollingStartIntervalNumber()
+                    it.rollingStartIntervalNumber >= dateWithMissingExposureKeys.toRollingStartIntervalNumber() &&
+                            it.rollingStartIntervalNumber <= dateWithMissingExposureKeys.endOfTheUtcDay()
+                        .toRollingIntervalNumber()
                 }
 
                 uploadTeksWithMessageType(keysToUpload, infectionLevel)
 
                 quarantineRepository.markMissingExposureKeysAsUploaded()
             } else {
                 // The regular flow of reporting the exposure keys.
-                val thresholdTime = if (infectionLevel == MessageType.InfectionLevel.Red) {
-                    diagnosisKeysRepository.getSentTeksByMessageType(
-                        MessageType.InfectionLevel.Yellow
-                    ).minBy { tekMetadata ->
-                        tekMetadata.rollingStartIntervalNumber
-                    }?.rollingStartIntervalNumber ?: thresholdTimeFromConfiguration
-                } else {
-                    thresholdTimeFromConfiguration
-                }
+                val uploadStartIntervalNumber =
+                    if (infectionLevel == MessageType.InfectionLevel.Red) {
+                        // If we sent a yellow warning before, send red warnings from the day of the
+                        // earliest yellow warning
+                        diagnosisKeysRepository.getSentTeksByMessageType(
+                            MessageType.InfectionLevel.Yellow
+                        ).map { tek ->
+                            tek.rollingStartIntervalNumber
+                        }.min() ?: uploadStartIntervalNumberFromConfig
+                    } else {
+                        uploadStartIntervalNumberFromConfig
+                    }
 
                 val teksToUpload = teks.filter {
```

Comment:
We might have an issue here, in case of going from yellow to red we are looking above for the oldest yellow rollingStartIntervalNumber from db which can be older than 14 days and then filter on the SDK teks which are only from the last 14 days. But maybe the teks for people that are older that 14 days are not important anymore because for those people the quarantine ended in the meantime.

Reply:
The argument goes exactly as you state. A 15 days old yellow key is no longer relevant and does not need to be revoked